File: frontend/src/services/admin/empresas.ts
import { GetSignal, POST, makeApiGetHandler } from "~/shared/http"
import { arrayToMapN } from "~/shared/main"

export interface IEmpresa {
  id: number,
  Email: string
  Nombre: string
  RazonSocial: string
  RUC: string
  Telefono: string
  Representante: string
  Direccion: string
  Ciudad: string
  SmtpConfig: IEmpresaSmtp
  CulquiConfig: IEmpresaCulqui
  ss: number
  upd: number
}

export interface IEmpresaSmtp {
  Host: string
  Port: string
  User: string
  Password: string
  Email: string
}

export interface IEmpresaCulqui {
  RsaKey: string
  RsaKeyID: string
  LlaveLive: string
  LlaveDev: string
  LlavePubLive: string
  LlavePubDev: string
}

export const useEmpresasAPI = (): GetSignal<IEmpresa[]> => {
  return  makeApiGetHandler(
    { route: "empresas", emptyValue: [],
      errorMessage: 'Hubo un error al obtener las empresas.',
      cacheSyncTime: 1, mergeRequest: true,
      useIndexDBCache: 'empresas',
    },
    (result) => {
      const empresa = result.Records.filter(x => x)[0] || {}
      empresa.smtp = empresa.smtp || {}
      return [empresa]
    }
  )
}

export const useParametrosEmpresaAPI = (): GetSignal<IEmpresa> => {
  return  makeApiGetHandler(
    { route: "empresa-parametros", emptyValue: [],
      errorMessage: 'Hubo un error al obtener las empresa.',
      cacheSyncTime: 1, // mergeRequest: true,
      useIndexDBCache: 'empresa',
    },
    (result) => {
      console.log("resultado obtenido:: ", result.Records)
      const empresa = (result.Records[0]|| {}) as IEmpresa
      empresa.SmtpConfig = empresa.SmtpConfig || {} as IEmpresaSmtp
      empresa.CulquiConfig = empresa.CulquiConfig || {} as IEmpresaCulqui

      return empresa
    }
  )
}

export interface IBackup {
  Name: string
  Size: number
  upd: number
}

export const useBackupsAPI = (): GetSignal<IBackup[]> => {
  return  makeApiGetHandler(
    { route: "backups", emptyValue: [],
      errorMessage: 'Hubo un error al obtener los backups.',
      cacheSyncTime: 0, mergeRequest: true,
      useIndexDBCache: 'backups',
    },
    ({ Records }) => {
      console.log("Backups obtenidos::", Records)
      return (Records || []) as IBackup[]
    }
  )
}

export const postEmpresaParametros = (data: IEmpresa) => {
  return POST({
    data,
    route: "empresa-parametros",
    refreshIndexDBCache: "empresa"
  })
}

export interface IUsuario {
  id: number,
  companyID: number
  nombres: string
  apellidos: string
  email: string
  usuario: string
  documentoNro: string
  cargo: string
  perfilesIDs: number[]
  rolesIDs: number[]
  ss: number
  upd: number
  created: number
  password1: string
  password2: string
}

export interface IUsuarioResult {
  usuarios: IUsuario[]
  usuariosMap: Map<number,IUsuario>
}

export const useUsuariosAPI = (): GetSignal<IUsuarioResult> => {
  return  makeApiGetHandler(
    { route: "usuarios", emptyValue: {},
      errorMessage: 'Hubo un error al obtener los usuarios.',
      cacheSyncTime: 1, mergeRequest: true,
      useIndexDBCache: 'usuarios',
    },
    (result) => {
      const usuarios = result.Records
      return {
        usuarios,
        usuariosMap: arrayToMapN(usuarios,'id')
      } as IUsuarioResult
    }
  )
}

export const postUsuario = (data: IUsuario) => {
  return POST({
    data,
    route: "usuarios",
    refreshIndexDBCache: "usuarios"
  })
}

export interface IAcceso {
  id: number
  nombre: string
  descripcion?: string
  grupo: number
  orden: number
  modulosIDs: number[]
  acciones: number[]
  ss: number
  upd: number
}

export const useAccesosAPI = (): GetSignal<IAcceso[]> => {
  return makeApiGetHandler(
    { route: "seguridad-accesos", emptyValue: [],
      errorMessage: 'Hubo un error al obtener los accesos.',
      cacheSyncTime: 1, mergeRequest: true,
      useIndexDBCache: 'seguridad_accesos',
    },
    (result) => {
      const accesos = result.Records
      console.log("accesos obtenidos::", accesos)
      return accesos
    }
  )
}

export const postSeguridadAccesos = (data: IAcceso) => {
  return POST({
    data,
    route: "seguridad-accesos",
    refreshIndexDBCache: "seguridad_accesos"
  })
}


export interface IPefil {
  id: number
  nombre: string
  descripcion?: string
  accesos: number[]
  modulosIDs: number[]
  accesosMap: Map<number,number[]>
  ss: number
  upd: number
  _open?: boolean
}

export interface IPerfiles {
  perfiles: IPefil[]
}

export const usePerfilesAPI = (): GetSignal<IPerfiles> => {
  return  makeApiGetHandler(
    { route: "perfiles", emptyValue: [],
      errorMessage: 'Hubo un error al obtener los perfiles.',
      cacheSyncTime: 1, mergeRequest: true, useIndexDBCache: 'perfiles',
    },
    (result) => {
      for(let pr of result.Records){
        pr.accesos = pr.accesos || []
        pr.accesosMap = pr.accesosMap || new Map()
        
        for(let e of pr.accesos){
          const id = Math.floor(e / 10)
          const nivel = e - (id * 10)
          pr.accesosMap.has(id) 
            ? pr.accesosMap.get(id).push(nivel)
            : pr.accesosMap.set(id, [nivel])
        }
      }
      
      const perfiles = result.Records.filter(x => x.ss > 0)
      console.log("Perfiles Obtenidos:: ", perfiles)
      return { perfiles }
    }
  )
}

export const postPerfil = (data: IPefil) => {
  return POST({
    data,
    route: "perfiles",
    refreshIndexDBCache: "perfiles"
  })
}

File: backend/configuracion/types/empresas.go
package types

import "app/db"

type TAGS struct{}

type Empresa struct {
	db.TableStruct[EmpresaTable, Empresa]
	ID                 int32        `db:"id,pk" col:",sk"`
	Nombre             string       `json:",omitempty" db:"nombre" col:""`
	RazonSocial        string       `json:",omitempty" db:"razon_social" col:""`
	RUC                string       `json:",omitempty" db:"ruc" col:",index"`
	Email              string       `json:",omitempty" db:"email" col:",index"`
	NotificacionEmail  string       `json:",omitempty" db:"notificacion_email" col:""`
	Telefono           string       `json:",omitempty" db:"telefono" col:""`
	Representante      string       `json:",omitempty" db:"representante" col:""`
	Direccion          string       `json:",omitempty" db:"direccion" col:""`
	Ciudad             string       `json:",omitempty" db:"ciudad" col:""`
	FormApiKey         string       `json:",omitempty" db:"form_api_key" col:""`
	EmailVerificado    int8         `json:",omitempty" db:"email_verificado" col:""`
	TelefonoVerificado int8         `json:",omitempty" db:"telefono_verificado" col:""`
	SmtpConfig         SmtpConfig   `json:",omitempty" db:"smtp_config" col:""`
	CulquiConfig       CulquiConfig `json:",omitempty" db:"culqui_config" col:""`
	Updated            int32        `json:"upd" db:"updated" col:",index"`
	Status             int8         `json:"ss" db:"status" col:""`
}

type EmpresaTable struct {
	db.TableStruct[EmpresaTable, Empresa]
	ID                 db.Col[EmpresaTable, int32]
	Nombre             db.Col[EmpresaTable, string]
	RazonSocial        db.Col[EmpresaTable, string]
	RUC                db.Col[EmpresaTable, string]
	Email              db.Col[EmpresaTable, string]
	NotificacionEmail  db.Col[EmpresaTable, string]
	Telefono           db.Col[EmpresaTable, string]
	Representante      db.Col[EmpresaTable, string]
	Direccion          db.Col[EmpresaTable, string]
	Ciudad             db.Col[EmpresaTable, string]
	FormApiKey         db.Col[EmpresaTable, string]
	EmailVerificado    db.Col[EmpresaTable, int8]
	TelefonoVerificado db.Col[EmpresaTable, int8]
	SmtpConfig         db.Col[EmpresaTable, SmtpConfig]
	CulquiConfig       db.Col[EmpresaTable, CulquiConfig]
	Updated            db.Col[EmpresaTable, int32]
	Status             db.Col[EmpresaTable, int8]
}

func (e EmpresaTable) GetSchema() db.TableSchema {
	return db.TableSchema{
		Name:         "empresas",
		UseSequences: true,
		Keys:         []db.Coln{e.ID.Autoincrement(0)},
	}
}

type SmtpConfig struct {
	Email    string `json:",omitempty"`
	User     string `json:",omitempty"`
	Password string `json:",omitempty"`
	Post     int32  `json:",omitempty"`
	Host     string `json:",omitempty"`
}

type CulquiConfig struct {
	RsaKey       string `json:",omitempty"`
	RsaKeyID     string `json:",omitempty"`
	LlaveLive    string `json:",omitempty"`
	LlavePubLive string `json:",omitempty"`
	LlaveDev     string `json:",omitempty"`
	LlavePubDev  string `json:",omitempty"`
}

type EmpresaPub struct {
	ID            int32  `json:"id"`
	Nombre        string `json:",omitempty"`
	CulqiRsaKey   string `json:",omitempty"`
	CulqiRsaKeyID string `json:",omitempty"`
	CulqiLlave    string `json:",omitempty"`
}
